modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,422 @@
+        doWrite(QUERY_READS,
+            () -> 1 + 16 + 8 + 8 + 8,
+            buf -> {
+                buf.put((byte)type.ordinal());
+                writeUuid(buf, queryNodeId);
+                buf.putLong(id);
+                buf.putLong(logicalReads);
+                buf.putLong(physicalReads);
+            });
+    }
+
+    /**
+     * @param sesId Session id.
+     * @param taskName Task name.
+     * @param startTime Start time in milliseconds.
+     * @param duration Duration.
+     * @param affPartId Affinity partition id.
+     */
+    public void task(IgniteUuid sesId, String taskName, long startTime, long duration, int affPartId) {
+        byte[] nameBytes = taskName.getBytes();
+
+        doWrite(TASK,
+            () -> 24 + 4 + nameBytes.length + 8 + 8 + 4,
+            buf -> {
+                writeIgniteUuid(buf, sesId);
+                buf.putInt(nameBytes.length);
+                buf.put(nameBytes);
+                buf.putLong(startTime);
+                buf.putLong(duration);
+                buf.putInt(affPartId);
+            });
+    }
+
+    /**
+     * @param sesId Session id.
+     * @param queuedTime Time job spent on waiting queue.
+     * @param startTime Start time in milliseconds.
+     * @param duration Job execution time.
+     * @param timedOut {@code True} if job is timed out.
+     */
+    public void job(IgniteUuid sesId, long queuedTime, long startTime, long duration, boolean timedOut) {
+        doWrite(JOB,
+            () -> 24 + 8 + 8 + 8 + 1,
+            buf -> {
+                writeIgniteUuid(buf, sesId);
+                buf.putLong(queuedTime);
+                buf.putLong(startTime);
+                buf.putLong(duration);
+                buf.put(timedOut ? (byte)1 : 0);
+            });
+    }
+
+    /**
+     * @param op Operation type.
+     * @param sizeSupplier Record size supplier.
+     * @param writer Record writer.
+     */
+    private void doWrite(OperationType op, IntSupplier sizeSupplier, Consumer<ByteBuffer> writer) {
+        int size = sizeSupplier.getAsInt() + /*type*/ 1;
+
+        SegmentedRingByteBuffer.WriteSegment seg = ringByteBuf.offer(size);
+
+        if (seg == null) {
+            if (smallBufLogged.compareAndSet(false, true)) {
+                log.warning("The performance statistics in-memory buffer size is too small. Some operations " +
+                    "will not be logged.");
+            }
+
+            return;
+        }
+
+        // Ring buffer closed (writer stopping) or maximum size reached.
+        if (seg.buffer() == null) {
+            seg.release();
+
+            if (!fileWriter.isCancelled() && stopByMaxSize.compareAndSet(false, true))
+                log.warning("The performance statistics file maximum size is reached.");
+
+            return;
+        }
+
+        ByteBuffer buf = seg.buffer();
+
+        buf.put((byte)op.ordinal());
```

Comment:
We should introduce an explicit index here.
Otherwise, any edit in the middle of `OperationType` will broke deserialization.

Reply:
Done.